### PR TITLE
Given the plugins a shape

### PR DESCRIPTION
**Pull request by szepeviktor**
- PSR-4
- PSR-12
- HookAnnotation

### CODE REVIEW

**Comment by curquiza** [comment]
Hello again @szepeviktor 
Remove the Draft PR status when you will need a review 🚀 We will try to be as reactive as possible, we are sorry if it takes time sometimes, it's often the rush for our team! 😄

**Comment by szepeviktor** [comment]
@timothyjensen Look! 👀 It is like a rebirth.

**Comment by szepeviktor** [comment]
Development continues 🚀 e.g. with a search for container ...

**Comment by szepeviktor** [comment]
@curquiza I was struggling with the injector for days. Now it is in place.

Could you read my code?

**Comment by szepeviktor** [comment]
Please be aware that I've already received feedback like

> Viktor, this is a WordPress site not a Symfony application

:)

**Comment by curquiza** [comment]
Hello @szepeviktor I'll let @eskombro check this PR since he was one of the authors of this repo.
He has a lot to do before so it might take time 🙂

**Comment by szepeviktor** [comment]
I just want a confirmation that this code meets your expectations.

**Review by eskombro — COMMENTED** [review]
Hey @szepeviktor! 

I am really excited about this huge improvement you are introducing to this plugin and repo, thanks a lot! The project structure and the code readability and in-code documentation are just great!

I checked what you are doing on this PR, but sadly I couldn't test it because it is not possible yet to activate the plugin on the wp-admin dashboard, I got a fatal error (or am I missing something?). In case you think I should be able to run it already, would you mind giving me an idea of the steps you follow to install / activate? 

I also got curious about your extension phpstan-wordpress. I see it has got some attention on GitHub, congrats!

My advice would be to try and finish a very basic plugin, something like an MVP that we can run and test, and after that introduce the changes gradually, instead of ending up with one single big change to switch it all at once, what do you think?

Anyways, amazing work, thanks again! 🎉

**Comment by szepeviktor** [comment]
Al right - all right.
I'm not a developer, I support development with technologies.

Running the plugin is the last step of me writing code.

**Comment by eskombro** [comment]
> Running the plugin is the last step of me writing code.

Oh, that's fine, I just wanted to be sure that I wasn't missing something :) 

> I was just afraid that OOP technologies won't be accepted in a WordPress plugin.

I don't see any reason why it wouldn't work, and I have seen a few examples (even some small articles talking about it like [this one](https://premium.wpmudev.org/blog/advanced-wordpress-development-writing-object-oriented-plugins/))but this is why I think it would be better to try and run a very basic version of it and make sure everything works together, and then add all the specific functionality, so in case there is a problem with the main plugin structure or some other issue we are not thinking about, we can realize that and correct it early enough!

**Comment by szepeviktor** [comment]
I am very 😢 sad, Samuel.

That article.

```php
function __construct( $config ) {
	$this->api_key = $config['api_key'];

	add_filter( 'the_content', array( $this, 'form' ) );
	add_action( 'wp_enqueue_scripts', array( $this, 'assets' ) );
	add_action( 'wp_ajax_msf_form_submit', array( $this, 'submissionHandler' ) );
	add_action( 'wp_ajax_nopriv_msf_form_submit', array( $this, 'submissionHandler' ) );
}
```

This is also procedural code masqueraded in a class.

Very sad. People are unable to think.

**Comment by szepeviktor** [comment]
> so in case there is a problem with the main plugin structure or some other issue we are not thinking about

Before writing code I've made sure I rule the plugin. All it functionality is in my hand.
That is prerequisite of sustainable software.

**Comment by eskombro** [comment]
> I am very 😢 sad, Samuel.
> That article.
> 
> This is also procedural code masqueraded in a class.

Don't be sad :D 

This article is indeed just a guide for people who want to transform their procedural way of coding plugins into OOP, so it is focused on that transition. But it just a simple example.

I don't see any reason to be worried about any OOP principles causing problems on Wordpress plugin development, I have seen polymorphism, inheritance, interfaces, association, abstraction... So it feels safe.

Also, don't be sad :)

**Comment by szepeviktor** [comment]
> do you have any specific concern?

Yes. Reader are made believe that putting functions into a class is called OOP.

An object is a sovereign entity able to do all of its duties without any outside help.

**Comment by eskombro** [comment]
> > do you have any specific concern?
> 
> Yes. Reader are made believe that putting functions into a class is called OOP.
> 
> An object is a sovereign entity able to do all of its duties without any outside help.

Hahaha, I agree, but let me rephrase:

Do you have any concern regarding the MeiliSearch plugin implementation? :)

**Comment by szepeviktor** [comment]
See the [immutable configuration class](https://github.com/meilisearch/meilisearch-wordpress/pull/20/files#diff-9727fe62ad60c6637a7def31b06f908a39f99f118a78280469d7a62eb67a9d68) that **is** OOP.

**Comment by szepeviktor** [comment]
> Do you have any concern regarding the MeiliSearch plugin implementation concerning OOP? :)

No, I strive to eliminate all traces of "WordPress programing" as much as I can.
The hook (filter/action) system is inherently procedural so we **will** have fake classes with hooked methods.

**Comment by curquiza** [comment]
Hello @szepeviktor!
Do you need help to finalize your PR? You can pass it as `Ready for review` when you think this is over 🙂

**Comment by szepeviktor** [comment]
Yes, I need to wake up!!

**Comment by curquiza** [comment]
Take your time, it's ok 🙂 I just needed to know!

**Comment by szepeviktor** [comment]
I have no schedule ❌📆
Everything is ASAP.

**Comment by szepeviktor** [comment]
Added missing bits and pieces.
Tomorrow I will run the code for the 1st time 🥇

**Comment by thijndehaas** [comment]
Should the post not be deleted when it is changes from published to draft?

```php
if ($post->post_status !== 'publish') {
    return $this->deletePostFromIndex($post->ID);
}

$this->indexPost($post);

**Comment by szepeviktor** [comment]
> Should the post not be deleted when it is changes from published to draft?

You are right.
I've sent this PR without knowing what it the goal of this plugin.
When I work I always make plans (docs) before writing code.

**Comment by curquiza** [comment]
Hello @szepeviktor and @thijndehaas 

FYI we chose to stop working on the meilisearch-wordpress plugin. It does not mean we will not continue the work in the future, but this is currently not the priority :)

Sorry for this, and thanks for your work and comments :)
